drivers/ipm/ipm_stm32_ipcc.c

```diff
@@ -0,0 +1,318 @@
+#define IPCC_EnableTransmitChannel(hipcc, ch)	\
+			LL_C2_IPCC_EnableTransmitChannel(hipcc, 1 << ch)
+#define IPCC_DisableReceiveChannel(hipcc, ch)	\
+			LL_C2_IPCC_DisableReceiveChannel(hipcc, 1 << ch)
+#define IPCC_DisableTransmitChannel(hipcc, ch)	\
+			LL_C2_IPCC_DisableTransmitChannel(hipcc, 1 << ch)
+
+#define IPCC_ClearFlag_CHx(hipcc, ch)	LL_C2_IPCC_ClearFlag_CHx(hipcc, 1 << ch)
+#define IPCC_SetFlag_CHx(hipcc, ch)	LL_C2_IPCC_SetFlag_CHx(hipcc, 1 << ch)
+
+#define IPCC_IsActiveFlag_CHx(hipcc, ch)	\
+			LL_C2_IPCC_IsActiveFlag_CHx(hipcc, 1 << ch)
+
+#define IPCC_ReadReg(hipcc, reg) READ_REG(hipcc->C2##reg)
+
+#define IPCC_ReadReg_SR(hipcc) READ_REG(hipcc->C2TOC1SR)
+#define IPCC_ReadOtherInstReg_SR(hipcc) READ_REG(hipcc->C1TOC2SR)
+
+#endif
+
+struct stm32_ipcc_mailbox_config {
+	void (*irq_config_func)(struct device *dev);
+	struct stm32_pclken pclken;
+};
+
+struct stm32_ipcc_mbx_data {
+	IPCC_TypeDef *hipcc;
+	u32_t num_ch;
+	ipm_callback_t callback;
+	void *callback_ctx;
+	struct device *clock;
+};
+
+static struct stm32_ipcc_mbx_data stm32_IPCC_data;
+
+static void stm32_ipcc_mailbox_rx_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct stm32_ipcc_mbx_data *data = DEV_DATA(dev);
+	unsigned int value = 0;
+	u32_t mask, i;
+
+	mask = (~IPCC_ReadReg(data->hipcc, MR)) & IPCC_ALL_MR_RXO_CH_MASK;
+	mask &= IPCC_ReadOtherInstReg_SR(data->hipcc) & IPCC_ALL_SR_CH_MASK;
+
+	for (i = 0; i < data->num_ch; i++) {
+		if (!((1 << i) & mask)) {
+			continue;
+		}
+		LOG_DBG("%s channel = %x\r\n", __func__, i);
+		/* mask the channel Free interrupt  */
+		IPCC_DisableReceiveChannel(data->hipcc, i);
+
+		if (data->callback) {
+			/* Only one MAILBOX, id is unused and set to 0 */
+			data->callback(data->callback_ctx, i, &value);
+		}
+		/* clear status to acknoledge message reception */
+		IPCC_ClearFlag_CHx(data->hipcc, i);
+		IPCC_EnableReceiveChannel(data->hipcc, i);
+	}
+}
+
+static void stm32_ipcc_mailbox_tx_isr(void *arg)
+{
+	struct device *dev = arg;
+	struct stm32_ipcc_mbx_data *data = DEV_DATA(dev);
+	u32_t mask, i;
+
+	mask = (~IPCC_ReadReg(data->hipcc, MR)) & IPCC_ALL_MR_TXF_CH_MASK;
+	mask = mask >> IPCC_C1MR_CH1FM_Pos;
+
+	mask &= IPCC_ReadReg_SR(data->hipcc) & IPCC_ALL_SR_CH_MASK;
+
+	for (i = 0; i <  data->num_ch; i++) {
+		if (!((1 << i) & mask)) {
+			continue;
+		}
+		LOG_DBG("%s channel = %x\r\n", __func__, i);
+		/* mask the channel Free interrupt */
+		IPCC_DisableTransmitChannel(data->hipcc, i);
+	}
+}
+
+static int stm32_ipcc_mailbox_ipm_send(struct device *d, int wait, u32_t id,
+				       const void *buff, int size)
+{
+	struct stm32_ipcc_mbx_data *data = d->driver_data;
+
+	ARG_UNUSED(wait);
+	ARG_UNUSED(buff);
+	ARG_UNUSED(size);
+
+	if (id >= data->num_ch) {
+		LOG_ERR("invalid id (%d)\r\n", id);
+		return  -EINVAL;
+	}
+
+	LOG_DBG("Send msg on channel %d\r\n", id);
+
+	/* Check that the channel is free (otherwise return error) */
+	if (IPCC_IsActiveFlag_CHx(data->hipcc, id)) {
```

Comment:
Comment doesn't match the code.
Also, we might want to implement the EBUSY API behavior